File: backend/src/models/Report.js
const mongoose = require('mongoose');
const { v4: uuidv4 } = require('uuid');

const reportSchema = new mongoose.Schema(
  {
    // Unique Report ID for tracking
    reportId: {
      type: String,
      unique: true,
      default: () => `RC-${Date.now().toString(36).toUpperCase()}-${uuidv4().split('-')[0].toUpperCase()}`,
    },
    
    // Reporter Information
    reporter: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null, // Null for anonymous reports
    },
    isAnonymous: {
      type: Boolean,
      default: false,
    },
    reporterEmail: {
      type: String, // For anonymous reporters who want updates
      trim: true,
    },
    
    // Incident Type and Category
    incidentType: {
      type: String,
      required: [true, 'Incident type is required'],
      enum: [
        'phishing_email',
        'phishing_sms',
        'phishing_social_media',
        'malware',
        'ransomware',
        'identity_theft',
        'online_fraud',
        'scam',
        'data_breach',
        'suspicious_website',
        'suspicious_link',
        'social_engineering',
        'cyberbullying',
        'harassment',
        'unauthorized_access',
        'ddos_attack',
        'cryptojacking',
        'other',
      ],
    },
    incidentSubtype: {
      type: String,
      trim: true,
    },
    
    // Incident Details
    title: {
      type: String,
      required: [true, 'Report title is required'],
      trim: true,
      maxlength: [200, 'Title cannot exceed 200 characters'],
    },
    description: {
      type: String,
      required: [true, 'Description is required'],
      maxlength: [10000, 'Description cannot exceed 10000 characters'],
    },
    
    // Temporal Information
    incidentDate: {
      type: Date,
      required: [true, 'Incident date is required'],
    },
    incidentTime: {
      type: String, // HH:MM format
    },
    
    // Severity
    severity: {
      type: String,
      enum: ['low', 'medium', 'high', 'critical'],
      default: 'medium',
    },
    
    // URLs and Links
    suspiciousUrls: [{
      url: {
        type: String,
        trim: true,
      },
      safetyCheck: {
        checked: { type: Boolean, default: false },
        isSafe: { type: Boolean, default: null },
        threats: [String],
        checkedAt: Date,
      },
    }],
    
    // Affected Platform/Service
    affectedPlatform: {
      type: String,
      trim: true,
    },
    affectedService: {
      type: String,
      trim: true,
    },
    
    // Geographic Information
    location: {
      country: String,
      region: String,
      city: String,
      coordinates: {
        latitude: Number,
        longitude: Number,
      },
      autoDetected: {
        type: Boolean,
        default: false,
      },
    },
    
    // Evidence/Attachments
    attachments: [{
      fileId: {
        type: String,
        required: true,
      },
      originalName: {
        type: String,
        required: true,
      },
      mimeType: {
        type: String,
        required: true,
      },
      size: {
        type: Number,
        required: true,
      },
      storageUrl: {
        type: String,
        required: true,
      },
      encryptionIv: {
        type: String,
      },
      fileHash: {
        type: String,
      },
      watermark: {
        type: mongoose.Schema.Types.Mixed,
      },
      malwareScan: {
        scanned: { type: Boolean, default: false },
        isSafe: { type: Boolean, default: null },
        scanResult: String,
        scannedAt: Date,
      },
      uploadedAt: {
        type: Date,
        default: Date.now,
      },
    }],
    
    // Report Status
    status: {
      type: String,
      enum: ['draft', 'submitted', 'under_review', 'verified', 'escalated', 'resolved', 'closed', 'rejected'],
      default: 'submitted',
    },
    statusHistory: [{
      status: String,
      changedBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
      changedAt: {
        type: Date,
        default: Date.now,
      },
      notes: String,
    }],
    
    // Priority (set by analysts/moderators)
    priority: {
      type: String,
      enum: ['low', 'normal', 'high', 'urgent'],
      default: 'normal',
    },
    
    // Assignment
    assignedTo: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
    },
    assignedAt: Date,
    
    // Verification
    verification: {
      isVerified: {
        type: Boolean,
        default: false,
      },
      verifiedBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
      verifiedAt: Date,
      verificationNotes: String,
    },
    
    // Evidence Verifications - References to Verification documents
    evidenceVerifications: [{
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Verification',
    }],
    
    // Tags for categorization
    tags: [{
      type: String,
      trim: true,
      lowercase: true,
    }],
    
    // Internal Notes (visible to analysts/moderators only)
    internalNotes: [{
      note: String,
      createdBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
      createdAt: {
        type: Date,
        default: Date.now,
      },
    }],
    
    // Public Comments
    comments: [{
      comment: String,
      createdBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
      isPublic: {
        type: Boolean,
        default: true,
      },
      createdAt: {
        type: Date,
        default: Date.now,
      },
    }],
    
    // Related Reports
    relatedReports: [{
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Report',
    }],
    
    // Escalation
    escalation: {
      isEscalated: {
        type: Boolean,
        default: false,
      },
      escalatedTo: String, // Authority name
      escalatedAt: Date,
      escalatedBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
      escalationReason: String,
    },
    
    // Resolution
    resolution: {
      isResolved: {
        type: Boolean,
        default: false,
      },
      resolvedAt: Date,
      resolvedBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
      resolutionSummary: String,
      outcome: {
        type: String,
        enum: ['confirmed_threat', 'false_positive', 'insufficient_evidence', 'duplicate', 'resolved_by_user', 'other'],
      },
    },
    
    // Auto-save Draft
    isDraft: {
      type: Boolean,
      default: false,
    },
    lastAutoSave: Date,
    
    // Consent
    dataProcessingConsent: {
      type: Boolean,
      required: [true, 'Data processing consent is required'],
    },
    
    // Analytics
    viewCount: {
      type: Number,
      default: 0,
    },
    
    // IP Address (for abuse prevention)
    submitterIp: {
      type: String,
      select: false,
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Indexes
reportSchema.index({ reportId: 1 });
reportSchema.index({ reporter: 1 });
reportSchema.index({ status: 1 });
reportSchema.index({ incidentType: 1 });
reportSchema.index({ severity: 1 });
reportSchema.index({ createdAt: -1 });
reportSchema.index({ incidentDate: -1 });
reportSchema.index({ tags: 1 });
reportSchema.index({ 'location.country': 1 });
reportSchema.index({ assignedTo: 1 });

// Text index for search
reportSchema.index({
  title: 'text',
  description: 'text',
  tags: 'text',
});

// Virtual for time since submission
reportSchema.virtual('timeSinceSubmission').get(function() {
  const now = new Date();
  const diff = now - this.createdAt;
  const days = Math.floor(diff / (1000 * 60 * 60 * 24));
  const hours = Math.floor((diff % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
  
  if (days > 0) {
    return `${days} day(s) ago`;
  }
  return `${hours} hour(s) ago`;
});

// Pre-save middleware to add status history
reportSchema.pre('save', function(next) {
  if (this.isModified('status') && !this.isNew) {
    this.statusHistory.push({
      status: this.status,
      changedAt: new Date(),
    });
  }
  next();
});

// Static method to get incident type categories
reportSchema.statics.getIncidentCategories = function() {
  return {
    phishing: ['phishing_email', 'phishing_sms', 'phishing_social_media'],
    malware: ['malware', 'ransomware'],
    fraud: ['identity_theft', 'online_fraud', 'scam'],
    security: ['data_breach', 'unauthorized_access', 'ddos_attack', 'cryptojacking'],
    suspicious: ['suspicious_website', 'suspicious_link', 'social_engineering'],
    harassment: ['cyberbullying', 'harassment'],
    other: ['other'],
  };
};

// Method to update status with history
reportSchema.methods.updateStatus = function(newStatus, userId, notes = '') {
  this.status = newStatus;
  this.statusHistory.push({
    status: newStatus,
    changedBy: userId,
    changedAt: new Date(),
    notes,
  });
};

const Report = mongoose.model('Report', reportSchema);

module.exports = Report;

File: frontend/lib/features/report/domain/entities/report.dart
import 'package:equatable/equatable.dart';

enum ReportType {
  phishing,
  malware,
  identityTheft,
  onlineFraud,
  dataBreach,
  ransomware,
  socialEngineering,
  ddos,
  cyberstalking,
  hacking,
  other,
}

enum ReportStatus {
  draft,
  submitted,
  underReview,
  investigating,
  resolved,
  closed,
  rejected,
}

enum SeverityLevel { low, medium, high, critical }

class Report extends Equatable {
  final String id;
  final String reportId;
  final ReportType type;
  final String title;
  final String description;
  final ReportStatus status;
  final SeverityLevel severity;
  final String? reporterId;
  final bool isAnonymous;
  final DateTime? incidentDate;
  final ReportLocation? location;
  final List<String> affectedSystems;
  final String? attackerInfo;
  final List<String> suspiciousUrls;
  final List<String> suspiciousEmails;
  final List<String> suspiciousIps;
  final double? financialLoss;
  final String? currency;
  final List<Evidence> evidence;
  final String? assignedTo;
  final int upvotes;
  final int downvotes;
  final bool hasUpvoted;
  final bool hasDownvoted;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Report({
    required this.id,
    required this.reportId,
    required this.type,
    required this.title,
    required this.description,
    required this.status,
    this.severity = SeverityLevel.medium,
    this.reporterId,
    this.isAnonymous = false,
    this.incidentDate,
    this.location,
    this.affectedSystems = const [],
    this.attackerInfo,
    this.suspiciousUrls = const [],
    this.suspiciousEmails = const [],
    this.suspiciousIps = const [],
    this.financialLoss,
    this.currency,
    this.evidence = const [],
    this.assignedTo,
    this.upvotes = 0,
    this.downvotes = 0,
    this.hasUpvoted = false,
    this.hasDownvoted = false,
    required this.createdAt,
    required this.updatedAt,
  });

  String get statusDisplayName {
    switch (status) {
      case ReportStatus.draft:
        return 'Draft';
      case ReportStatus.submitted:
        return 'Submitted';
      case ReportStatus.underReview:
        return 'Under Review';
      case ReportStatus.investigating:
        return 'Investigating';
      case ReportStatus.resolved:
        return 'Resolved';
      case ReportStatus.closed:
        return 'Closed';
      case ReportStatus.rejected:
        return 'Rejected';
    }
  }

  String get typeDisplayName {
    switch (type) {
      case ReportType.phishing:
        return 'Phishing';
      case ReportType.malware:
        return 'Malware';
      case ReportType.identityTheft:
        return 'Identity Theft';
      case ReportType.onlineFraud:
        return 'Online Fraud';
      case ReportType.dataBreach:
        return 'Data Breach';
      case ReportType.ransomware:
        return 'Ransomware';
      case ReportType.socialEngineering:
        return 'Social Engineering';
      case ReportType.ddos:
        return 'DDoS Attack';
      case ReportType.cyberstalking:
        return 'Cyberstalking';
      case ReportType.hacking:
        return 'Hacking';
      case ReportType.other:
        return 'Other';
    }
  }

  Report copyWith({
    String? id,
    String? reportId,
    ReportType? type,
    String? title,
    String? description,
    ReportStatus? status,
    SeverityLevel? severity,
    String? reporterId,
    bool? isAnonymous,
    DateTime? incidentDate,
    ReportLocation? location,
    List<String>? affectedSystems,
    String? attackerInfo,
    List<String>? suspiciousUrls,
    List<String>? suspiciousEmails,
    List<String>? suspiciousIps,
    double? financialLoss,
    String? currency,
    List<Evidence>? evidence,
    String? assignedTo,
    int? upvotes,
    int? downvotes,
    bool? hasUpvoted,
    bool? hasDownvoted,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Report(
      id: id ?? this.id,
      reportId: reportId ?? this.reportId,
      type: type ?? this.type,
      title: title ?? this.title,
      description: description ?? this.description,
      status: status ?? this.status,
      severity: severity ?? this.severity,
      reporterId: reporterId ?? this.reporterId,
      isAnonymous: isAnonymous ?? this.isAnonymous,
      incidentDate: incidentDate ?? this.incidentDate,
      location: location ?? this.location,
      affectedSystems: affectedSystems ?? this.affectedSystems,
      attackerInfo: attackerInfo ?? this.attackerInfo,
      suspiciousUrls: suspiciousUrls ?? this.suspiciousUrls,
      suspiciousEmails: suspiciousEmails ?? this.suspiciousEmails,
      suspiciousIps: suspiciousIps ?? this.suspiciousIps,
      financialLoss: financialLoss ?? this.financialLoss,
      currency: currency ?? this.currency,
      evidence: evidence ?? this.evidence,
      assignedTo: assignedTo ?? this.assignedTo,
      upvotes: upvotes ?? this.upvotes,
      downvotes: downvotes ?? this.downvotes,
      hasUpvoted: hasUpvoted ?? this.hasUpvoted,
      hasDownvoted: hasDownvoted ?? this.hasDownvoted,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      id: json['_id'] ?? json['id'],
      reportId: json['reportId'] ?? '',
      type: _parseReportType(json['type']),
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      status: _parseReportStatus(json['status']),
      severity: _parseSeverity(json['severity']),
      reporterId: json['reporter'],
      isAnonymous: json['isAnonymous'] ?? false,
      incidentDate: json['incidentDate'] != null
          ? DateTime.parse(json['incidentDate'])
          : null,
      location: json['location'] != null
          ? ReportLocation.fromJson(json['location'])
          : null,
      affectedSystems: json['affectedSystems'] != null
          ? List<String>.from(json['affectedSystems'])
          : [],
      attackerInfo: json['attackerInfo'],
      suspiciousUrls: json['suspiciousUrls'] != null
          ? List<String>.from(json['suspiciousUrls'])
          : [],
      suspiciousEmails: json['suspiciousEmails'] != null
          ? List<String>.from(json['suspiciousEmails'])
          : [],
      suspiciousIps: json['suspiciousIps'] != null
          ? List<String>.from(json['suspiciousIps'])
          : [],
      financialLoss: json['financialLoss']?.toDouble(),
      currency: json['currency'],
      evidence: json['evidence'] != null
          ? (json['evidence'] as List).map((e) => Evidence.fromJson(e)).toList()
          : [],
      assignedTo: json['assignedTo'],
      upvotes: json['upvotes'] ?? 0,
      downvotes: json['downvotes'] ?? 0,
      hasUpvoted: json['hasUpvoted'] ?? false,
      hasDownvoted: json['hasDownvoted'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'reportId': reportId,
        'type': type.name,
        'title': title,
        'description': description,
        'status': status.name,
        'severity': severity.name,
        'reporter': reporterId,
        'isAnonymous': isAnonymous,
        'incidentDate': incidentDate?.toIso8601String(),
        'location': location?.toJson(),
        'affectedSystems': affectedSystems,
        'attackerInfo': attackerInfo,
        'suspiciousUrls': suspiciousUrls,
        'suspiciousEmails': suspiciousEmails,
        'suspiciousIps': suspiciousIps,
        'financialLoss': financialLoss,
        'currency': currency,
        'evidence': evidence.map((e) => e.toJson()).toList(),
        'assignedTo': assignedTo,
        'upvotes': upvotes,
        'downvotes': downvotes,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };

  static ReportType _parseReportType(String? type) {
    switch (type?.toLowerCase()) {
      case 'phishing':
        return ReportType.phishing;
      case 'malware':
        return ReportType.malware;
      case 'identity_theft':
      case 'identitytheft':
        return ReportType.identityTheft;
      case 'online_fraud':
      case 'onlinefraud':
        return ReportType.onlineFraud;
      case 'data_breach':
      case 'databreach':
        return ReportType.dataBreach;
      case 'ransomware':
        return ReportType.ransomware;
      case 'social_engineering':
      case 'socialengineering':
        return ReportType.socialEngineering;
      case 'ddos':
        return ReportType.ddos;
      default:
        return ReportType.other;
    }
  }

  static ReportStatus _parseReportStatus(String? status) {
    switch (status?.toLowerCase()) {
      case 'draft':
        return ReportStatus.draft;
      case 'submitted':
        return ReportStatus.submitted;
      case 'under_review':
      case 'underreview':
        return ReportStatus.underReview;
      case 'investigating':
        return ReportStatus.investigating;
      case 'resolved':
        return ReportStatus.resolved;
      case 'closed':
        return ReportStatus.closed;
      case 'rejected':
        return ReportStatus.rejected;
      default:
        return ReportStatus.draft;
    }
  }

  static SeverityLevel _parseSeverity(String? severity) {
    switch (severity?.toLowerCase()) {
      case 'low':
        return SeverityLevel.low;
      case 'medium':
        return SeverityLevel.medium;
      case 'high':
        return SeverityLevel.high;
      case 'critical':
        return SeverityLevel.critical;
      default:
        return SeverityLevel.medium;
    }
  }

  @override
  List<Object?> get props => [
        id,
        reportId,
        type,
        title,
        description,
        status,
        severity,
        reporterId,
        isAnonymous,
        incidentDate,
        location,
        affectedSystems,
        evidence,
        upvotes,
        downvotes,
        createdAt,
        updatedAt,
      ];
}

class ReportLocation extends Equatable {
  final String? country;
  final String? state;
  final String? city;
  final double? latitude;
  final double? longitude;

  const ReportLocation({
    this.country,
    this.state,
    this.city,
    this.latitude,
    this.longitude,
  });

  factory ReportLocation.fromJson(Map<String, dynamic> json) {
    return ReportLocation(
      country: json['country'],
      state: json['state'],
      city: json['city'],
      latitude: json['coordinates']?['latitude']?.toDouble(),
      longitude: json['coordinates']?['longitude']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
        'country': country,
        'state': state,
        'city': city,
        if (latitude != null && longitude != null)
          'coordinates': {'latitude': latitude, 'longitude': longitude},
      };

  @override
  List<Object?> get props => [country, state, city, latitude, longitude];
}

class Evidence extends Equatable {
  final String id;
  final String fileName;
  final String fileType;
  final int fileSize;
  final String url;
  final String? thumbnailUrl;
  final bool isEncrypted;
  final DateTime uploadedAt;

  const Evidence({
    required this.id,
    required this.fileName,
    required this.fileType,
    required this.fileSize,
    required this.url,
    this.thumbnailUrl,
    this.isEncrypted = true,
    required this.uploadedAt,
  });

  factory Evidence.fromJson(Map<String, dynamic> json) {
    return Evidence(
      id: json['_id'] ?? json['id'] ?? '',
      fileName: json['fileName'] ?? json['originalName'] ?? '',
      fileType: json['fileType'] ?? json['mimeType'] ?? '',
      fileSize: json['fileSize'] ?? 0,
      url: json['url'] ?? json['firebaseUrl'] ?? '',
      thumbnailUrl: json['thumbnailUrl'],
      isEncrypted: json['isEncrypted'] ?? true,
      uploadedAt: json['uploadedAt'] != null
          ? DateTime.parse(json['uploadedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'fileName': fileName,
        'fileType': fileType,
        'fileSize': fileSize,
        'url': url,
        'thumbnailUrl': thumbnailUrl,
        'isEncrypted': isEncrypted,
        'uploadedAt': uploadedAt.toIso8601String(),
      };

  @override
  List<Object?> get props => [id, fileName, fileType, url, uploadedAt];
}
